src/vsphere_cpi/lib/ruby_vim_sdk/vm_power_on_exception.rb

```diff
@@ -0,0 +1,12 @@
+module VimSdk
+  class VMPowerOnError < StandardError
+
+    attr_reader :fault
+
+    def initialize(message, fault)
```

Comment:
The `raise` method in Ruby is really strange in that it takes between 0 and 3 arguments and doesn't directly create an exception:

- 0 arguments: raises whatever is in `$!` (almost always the current exception being handled).
- 1 argument: if the argument responds to the `exception` method then it calls `exception` on the argument and raises that. Both subclasses (e.g. `TypeError`) and instances (e.g. `TypeError.new`) of `Exception` respond to the `exception` method. If the argument is a `String` then this raises a `RuntimeError` with the argument as the *message*.
- 2 arguments: raises the result of calling `exception` on the first argument with the second argument as the message.
- 3 arguments: raises the result of calling `exception` on the first argument with the second argument as the message and the third argument as the backtrace.

It's stylistically preferable in Ruby to use the 2 argument form of `raise` to supply a message rather than the 1 argument form (see https://github.com/bbatsov/ruby-style-guide#exception-class-messages). Therefore we can define the `initialize` method instead as:

```ruby
def initialize(fault)
  @fault = fault
end
```

So that later on we can raise this exception as:

```ruby
raise VimSdk::VMPowerOnError.new(fault), "Could not power on VM '#{vm}': #{fault.msg}"
```

You can probably go further here and define `to_s` as:

```ruby
def to_s
  "#{super}: #{fault.msg}"
end
```

So that the fault message is automatically appended and your `raise`s become:

```ruby
raise VimSdk::VMPowerOnError.new(fault), "Could not power on VM '#{vm}'"
```

Reply:
Fixed